src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -1347,25 +1353,60 @@ QueryPlanPtr MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreamsFinal(
         {
             Pipes pipes;
 
-            for (auto part_it = parts_to_merge_ranges[range_index]; part_it != parts_to_merge_ranges[range_index + 1]; ++part_it)
+            /// If do_not_merge_across_partitions_select_final is true and there is only one part in partition
+            /// with level > 0 then we won't postprocess this part and if num_streams > 1 we
+            /// can use parallel select on this part.
+            if (num_streams > 1 && settings.do_not_merge_across_partitions_select_final &&
+                std::distance(parts_to_merge_ranges[range_index], parts_to_merge_ranges[range_index + 1]) == 1 &&
+                parts_to_merge_ranges[range_index]->data_part->info.level > 0)
             {
-                auto source_processor = std::make_shared<MergeTreeSelectProcessor>(
+                MergeTreeReadPoolPtr pool = std::make_shared<MergeTreeReadPool>(
+                    num_streams,
+                    sum_marks,
+                    min_marks_for_concurrent_read,
+                    std::vector{*std::move(parts_to_merge_ranges[range_index])},
                     data,
                     metadata_snapshot,
-                    part_it->data_part,
-                    max_block_size,
-                    settings.preferred_block_size_bytes,
-                    settings.preferred_max_column_in_block_size_bytes,
-                    column_names,
-                    part_it->ranges,
-                    use_uncompressed_cache,
                     query_info.prewhere_info,
                     true,
-                    reader_settings,
-                    virt_columns,
-                    part_it->part_index_in_query);
+                    column_names,
+                    MergeTreeReadPool::BackoffSettings(settings),
+                    settings.preferred_block_size_bytes,
+                    false);
+
```

Comment:
Worth adding the same message as in https://github.com/ClickHouse/ClickHouse/blob/cc11e7c2f2ea0c508c30748623d51c176a5ad21b/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp#L987 ?

And also count "approx" rows?